### PR TITLE
load: fix load data with non-utf8 will succeed

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25979

### What is changed and how it works?

How it Works: 

Refer: https://dev.mysql.com/doc/refman/8.0/en/load-data.html
**With IGNORE or LOCAL, data-interpretation errors become warnings and the load operation continues, even if the SQL mode is restrictive. For examples, see Column Value Assignment.**

Reason:
load data into t:  CAST(BINARY(X'c2c2c2c2') -> CHAR(2); 
the returned truncated value is X'c2c2' which is still invalid without right utf8 encode.

**That's is because the loaddata statement's TruncateAsWarning doesn't follow the strict sql mode when keyword `ignore` and `local` is turned on.**

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
```
Using the real TiKV nodes with TiUP

create table t( `a` varchar(12) DEFAULT NUL);
load data local infile "/home/arenatlx/下载/1111.txt" into table t FIELDS TERMINATED BY '||' LINES TERMINATED BY '\n' (@`a`) set `a` = NULLif(@a,'');

analyze table t // shouldn't have a error here.

``` 
[1111.txt](https://github.com/pingcap/tidb/files/6783451/1111.1.txt)

### Release note 

- load: fix load data with non-utf8 can succeed

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -1744,7 +1744,13 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	case *ast.LoadDataStmt:
 		sc.DupKeyAsWarning = true
 		sc.BadNullAsWarning = true
-		sc.TruncateAsWarning = !vars.StrictSQLMode
+		// With IGNORE or LOCAL, data-interpretation errors become warnings and the load operation continues,
+		// even if the SQL mode is restrictive. For examples, see Column Value Assignment.
+		if stmt.IsLocal || stmt.OnDuplicate == ast.OnDuplicateKeyHandlingIgnore {
+			sc.TruncateAsWarning = true
+		} else {
+			sc.TruncateAsWarning = !vars.StrictSQLMode
```
Comment (wjhuang2016) [inline executor/executor.go:1752]: In what case can TiDB reach this line？

Comment (AilinKid) [reply executor/executor.go:1752]: eg: load data without ignore keyword and local keyword

Comment (jackysp) [reply executor/executor.go:1752]: I think `load data` without `local` is not supported in TiDB.

Comment (AilinKid) [reply executor/executor.go:1752]: yes

**Comment by AilinKid** [comment]
PTAL @wjhuang2016 @jackysp

File: executor/executor.go (hunk 1 above)
Comment (wjhuang2016) [reply executor/executor.go:1752]: @jackysp means that `sc.TruncateAsWarning = true` , no if.

File: executor/executor.go [hunk 2]
```diff
@@ -1745,7 +1745,13 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	case *ast.LoadDataStmt:
 		sc.DupKeyAsWarning = true
 		sc.BadNullAsWarning = true
-		sc.TruncateAsWarning = !vars.StrictSQLMode
+		// With IGNORE or LOCAL, data-interpretation errors become warnings and the load operation continues,
```
Comment (bb7133) [inline executor/executor.go:1748]: Could you add a test case for the change?

File: executor/executor.go (hunk 1 above)
Comment (AilinKid) [reply executor/executor.go:1753]: sure

File: executor/executor.go (hunk 2 above)
Comment (AilinKid) [reply executor/executor.go:1748]: we got the manual test scripts

File: executor/executor.go [hunk 3]
```diff
@@ -1745,7 +1745,13 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
...
+		// even if the SQL mode is restrictive. For examples, see Column Value Assignment.
```
Comment (tangenta) [inline executor/executor.go:1749]: What is "Column Value Assignment"? Is there any link to the details?

**Review by wjhuang2016 — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**